haystack/nodes/prompt/providers.py

```diff
@@ -402,10 +411,32 @@ def invoke(self, *args, **kwargs):
             "best_of": kwargs_with_defaults.get("best_of", 1),
             "logit_bias": kwargs_with_defaults.get("logit_bias", {}),
         }
-        res = openai_request(url=self.url, headers=self.headers, payload=payload)
-        _check_openai_text_completion_answers(result=res, payload=payload)
-        responses = [ans["text"].strip() for ans in res["choices"]]
-        return responses
+        if not stream:
+            res = openai_request(url=self.url, headers=self.headers, payload=payload)
+            _check_openai_text_completion_answers(result=res, payload=payload)
+            responses = [ans["text"].strip() for ans in res["choices"]]
+            return responses
+        else:
+            response = openai_request(
+                url=self.url, headers=self.headers, payload=payload, load_response=False, stream=True
+            )
+
+            def default_stream_handler(token_received: str) -> str:
+                print(token_received, flush=True, end="")
+                return token_received
+
+            handler: Callable[[str], str] = kwargs_with_defaults.pop("stream_handler", default_stream_handler)
+            client = sseclient.SSEClient(response)
+            tokens: List[str] = []
+            try:
+                for event in client.events():
+                    if event.data != "[DONE]":
+                        ed = json.loads(event.data)
+                        token: str = ed["choices"][0]["text"]
+                        tokens.append(handler(token))
+            finally:
+                client.close()
+            return ["".join(tokens)]  # return a list of strings just like non-streaming
```

Comment:
I'd agree with you here, but no returning response here as we do on line 439 has wide-ranging consequences. Most PromptNode code relies on synchronous results returned from PromptNode. If people want to use streaming, it would all break and require a code revamp. The approach of returning `["".join(tokens)]` enables the use of streaming (for UX purposes mainly), yet no code change is required for PromptNode users. They may not need to print the result twice. But that's an easy fix.  Please lmk know your thoughts @ZanSara

Reply:
Aah ok I see the issue. No problem then, it's not such a big deal. Let's keep it like this for now :+1: